Document how to use the automatic tuning featore of the pid component

petterreinholdtsen opened the pull request:
Include script to help do the tuning.

Fixes #2117

PetteriAimonen commented on docs/src/motion/pid-theory.adoc:
```diff
@@ -188,4 +188,98 @@ After tuning the axis check the following error with Halscope to make
+'FF1' while setting 'Dgain', 'FF0' and 'FF2' to zero.  Note type 1
+require scaling be set so output is in user units per second.
+
+When autotuning a motor with tupe-type 0, the algorithm will produce a
```
Minor typo: `tupe-type`

PetteriAimonen commented on docs/src/motion/pid-theory.adoc:
```diff
@@ -188,4 +188,98 @@ After tuning the axis check the following error with Halscope to make
+point for further parameter tuning.
+
+FIXME: The author of these instructions have not tested automatic
+tuning with type-type set to 1, so this approach remain to be
```
`type-type`?

PetteriAimonen commented on docs/src/motion/pid-theory.adoc:
```diff
@@ -188,4 +188,98 @@ After tuning the axis check the following error with Halscope to make
+When autotuning a motor with tupe-type 0, the algorithm will produce a
+square wave pattern centered around the 'bias' value on the output pin
+of the PID controller, moving from the positive extreme to the
+negative extreme of the output range.  This can be seen using the HAL
```
Might be good to mention the `ferror` limits here also, even though the script does extend them automatically. The `ferror` can also help to automatically stop if the PID tuning goes way out of range, to prevent hitting any endstops.

PetteriAimonen commented on scripts/run-auto-pid-tuner:
```diff
@@ -0,0 +1,155 @@
+# Some values that need to be replaced during tuning
+replace = {
+#    f"pid.{axis}.tune-cycles": "0x40",  # Set number of back and forth movement
+    f"pid.{axis}.tune-effort": "0.2",  # Set effort, should start low
```
Why `tune-cycles` is commented and in hex?
Is `tune-effort` in PID output pin units or something else?

petterreinholdtsen commented:
Thank you very much for the feedback and proof reading.

[Petteri Aimonen]
> Minor typo: `tupe-type`

Fixed.

> `type-type`?

'tune-type'.  Fixed.

> Might be good to mention the `ferror` limits here also, even though
> the script does extend them automatically. The `ferror` can also help
> to automatically stop if the PID tuning goes way out of range, to
> prevent hitting any endstops.

Added info on ferror.

> Why `tune-cycles` is commented and in hex?

It just document how to increase the number of cycles.  Adjusted comment
to make it clear.

> Is `tune-effort` in PID output pin units or something else?

Yes.  Adjusted pid(9) manual page to make this more clear.

-- 
Happy hacking
Petter Reinholdtsen

hansu commented on docs/src/motion/pid-theory.adoc:
```diff
@@ -188,4 +188,102 @@ After tuning the axis check the following error with Halscope to make
+its control signal, this might accellerate the motor full speed in one
+direction for a short period before telling it to go full speed in the
+opposite direction.  Make sure to have a lot of room on either side of
+the starting position, and start with a low `tune-effort` value to
+limit the speed used.  Using too high 'tune-effort' might overload the
```
What is the correlation between tune-effort and controller output signal?
Does tune-effort ? 0.5 limit the output to +-5V in the above example? If yes, would be good to mention.

petterreinholdtsen replied:
It is one to one.  If tune-effort is set to 0.5, the output will move from 0.5 to -0.5 (ie +-0.5V).  Will try to make it more clear.

hansu reviewed (commented):
Only some little things, other than that very helpful instructions :+1:

petterreinholdtsen commented on docs/src/motion/pid-theory.adoc:
```diff
@@ -188,4 +188,102 @@ After tuning the axis check the following error with Halscope to make
+to make sure it do not hit anything when it start moving back and
+forth.  You also want to extend the axis ferror limit (following
+error) to make LinuxCNC accept the wanted position to deviate from the
+actual position.  Sensible limits depend on the machine and setup, but
+1 inch or 20 mm miht be a useful starting points.  Next, set the
```
Yes, ferror is ment.  I'll add it to the text.

hansu replied:
Ok. But is it needed to set ferror that high? I remember that even setting ferror from 0.01 to 0.1 allowed to run auto tune on your machine if I remember correctly.

petterreinholdtsen replied:
I do not know how high it need to be.  I guess it depend on the machine.

petterreinholdtsen commented:
All suggestions have been integrated into the rebased and updated pull request.

hansu commented on docs/src/motion/pid-theory.adoc:
```diff
@@ -187,4 +187,106 @@ After tuning the axis check the following error with Halscope to make
+to make sure it do not hit anything when it start moving back and
+forth.  You also want to extend the axis ferror limit (following
+error) to make LinuxCNC accept the higher position deviation during
+tuning.  Sensible ferror limit depend on the machine and setup, but
+1 inch or 20 mm might be a useful starting points.  Next, set the
```
```suggestion
tuning.  The sensible ferror limit depends on the machine and setup, but
1 inch or 20 mm might be useful starting points.  Next, set the
```

petterreinholdtsen replied:
Look like a better phrase.
